packages/mui-base/src/Select/Option/SelectOption.tsx

```diff
@@ -0,0 +1,334 @@
+  // │ These PropTypes are generated from the TypeScript type definitions. │
+  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
+  // └─────────────────────────────────────────────────────────────────────┘
+  /**
+   * @ignore
+   */
+  children: PropTypes.node,
+  /**
+   * Class names applied to the element or a function that returns them based on the component's state.
+   */
+  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
+  /**
+   * If `true`, the select option will be disabled.
+   * @default false
+   */
+  disabled: PropTypes.bool,
+  /**
+   * @ignore
+   */
+  getRootItemProps: PropTypes.func.isRequired,
+  /**
+   * @ignore
+   */
+  highlighted: PropTypes.bool.isRequired,
+  /**
+   * @ignore
+   */
+  indexRef: PropTypes.shape({
+    current: PropTypes.number.isRequired,
+  }).isRequired,
+  /**
+   * A text representation of the select option's content.
+   * Used for keyboard text navigation matching.
+   */
+  label: PropTypes.string,
+  /**
+   * @ignore
+   */
+  open: PropTypes.bool.isRequired,
+  /**
+   * @ignore
+   */
+  popupRef: PropTypes.shape({
+    current: PropTypes.object,
+  }).isRequired,
+  /**
+   * A function to customize rendering of the component.
+   */
+  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
+  /**
+   * @ignore
+   */
+  selected: PropTypes.bool.isRequired,
+  /**
+   * @ignore
+   */
+  selectedIndexRef: PropTypes.shape({
+    current: PropTypes.number,
+  }).isRequired,
+  /**
+   * @ignore
+   */
+  selectionRef: PropTypes.shape({
+    current: PropTypes.shape({
+      allowSelect: PropTypes.bool.isRequired,
+      allowSelectedMouseUp: PropTypes.bool.isRequired,
+      allowUnselectedMouseUp: PropTypes.bool.isRequired,
+    }).isRequired,
+  }).isRequired,
+  /**
+   * @ignore
+   */
+  setActiveIndex: PropTypes.func.isRequired,
+  /**
+   * @ignore
+   */
+  setOpen: PropTypes.func.isRequired,
+  /**
+   * @ignore
+   */
+  setValue: PropTypes.func.isRequired,
+  /**
+   * @ignore
+   */
+  typingRef: PropTypes.shape({
+    current: PropTypes.bool.isRequired,
+  }).isRequired,
+  /**
+   * @ignore
+   */
+  value: PropTypes.any.isRequired,
+} as any;
+
+const MemoizedInnerSelectOption = React.memo(InnerSelectOption);
+
+/**
+ *
+ * Demos:
+ *
+ * - [Select](https://base-ui.netlify.app/components/react-select/)
+ *
+ * API:
+ *
+ * - [SelectOption API](https://base-ui.netlify.app/components/react-select/#api-reference-SelectOption)
+ */
+const SelectOption = React.forwardRef(function SelectOption(
+  props: SelectOption.Props,
+  forwardedRef: React.ForwardedRef<HTMLDivElement>,
+) {
+  const { value: valueProp = null, label, ...otherProps } = props;
+
+  const listItem = useCompositeListItem({ label });
+  const { activeIndex, selectedIndex, setActiveIndex } = useSelectIndexContext();
+  const { getItemProps, setOpen, setValue, open, selectionRef, typingRef, valuesRef, popupRef } =
```

Comment:
Can this be moved to InnerSelectOption? It seems that these values don't change on each option highlight (except maybe `getItemProps`?)

Reply:
Looks like you may be right (`getItemProps` also changes infrequently). I'll try to refactor this.

Reply:
Consuming the root context inside the memoized component is risky as any time the root context changes it will cause every option to re-render. Even if they don't change on each option highlight, they can change on other unrelated state changes. Minor refactorings could introduce issues unknowingly as well.